I18n support proposition

agsoto opened the pull request:
Hello! This is the draft we talked about on #82. Key considerations I had for the changes:

- Keeping components code under packages to be reusable
- Maintaining configuration consistent with both site and Astro config files
- Trying to make sure the update won't introduce any breaking changes  
- Providing a ready to use component approach for multilingual websites that appears only if there's i18n configuration

Caveats:
- Files under [lang] could be provided ready to use or just documented examples of how to use the features
- There are more file changes than I expected but many of them are just function calls 
- There are still some chores and to-dos such as moving a few functions logic, unify lang property on markdown files, using translation functions or change components imports if those are merged
- Excuses in advance if there's any error on Zh translations! Those were created with online tools

I understand you're busy right now but I would like to know your thoughts on the approach when you have the time. 
Thank you! 😃

agsoto committed:
feat(i18n): add locale-based routing and language switch

- add src/pages/[lang]/** routes for localized blog and pages
- add packages/pure/components/basic/LanguageSwitch.astro and LocaleRedirect.astro
- add locale dictionaries packages/pure/i18n/locales/{en,es,zh}/common.ts
- add locale utils packages/pure/utils/locale.ts and locale-client-redirect.ts
- add src/middleware.ts for locale detection and redirects
- update header/footer, layouts, TOC, and post preview to use i18n
- update astro.config.ts , src/site.config.ts , and tsconfig.json for locales
- move zh blog content to src/content/localized/zh/blog/**
- update exports and types in packages/pure/index.ts , components/basic/index.ts , types/* , and plugins/virtual-user-config.ts

agsoto committed:
chore(content): move zh blog content to localized folder and remove old path

Delete src/content/blog/markdown-zh/* and add localized copies under src/content/localized/zh/blog/markdown-zh

cworld1 commented on src/pages/[lang]/blog/[...id].astro:
```diff
@@ -0,0 +1,37 @@
+import { type CollectionEntry, getCollection, getEntry, render } from 'astro:content'
+import PostLayout from '@/layouts/BlogPost.astro'
+import { getRelativeLocaleUrlList } from 'astro:i18n'
+import { normalizeLocaleBlogId } from '../../../../packages/pure/utils/locale'
```
'../../../../packages/pure/utils/locale' is a strange writing method. you can use `astro-pure/utils` as if [package is connected](https://bun.com/docs/pm/cli/link) and function is exported in utils default file

agsoto commented:
I tried to replicate the error on my machine changing configuration such as the Astro routing with prefixDefaultLocale: true and disabling redirections but still no similar error appears. Do you get the error on the home?

agsoto committed:
chore(pages): translate strings for localized blog page

cworld1 commented:
I've never tried `prefixDefaultLocale` before. OK, the deployment looks good, with some warnings: 

```
09:44:54   └─ /archives/index.html09:44:54 [WARN] `Astro.request.headers` was used when rendering the route `src/pages/archives/index.astro'`. `Astro.request.headers` is not available on prerendered pages. If you need access to request headers, make sure that the page is server-rendered using `export const prerender = false;` or by setting `output` to `"server"` in your Astro config to make all your pages server-rendered by default.
```

Can you making sure your code can work well with static mode? I mean, build it, and make a server like python http to test it

agsoto commented:
> I've never tried `prefixDefaultLocale` before. OK, the deployment looks good, with some warnings:
> 
> ```
> 09:44:54   └─ /archives/index.html09:44:54 [WARN] `Astro.request.headers` was used when rendering the route `src/pages/archives/index.astro'`. `Astro.request.headers` is not available on prerendered pages. If you need access to request headers, make sure that the page is server-rendered using `export const prerender = false;` or by setting `output` to `"server"` in your Astro config to make all your pages server-rendered by default.
> ```
> 
> Can you making sure your code can work well with static mode? I mean, build it, and make a server like python http to test it

Sure, I'll do it in a new github repo just for dev-test purposes.

> Seems something not work well? I checked the deployment, something mixed the zh and es language, and seems cannot render content when the language is not the default ones. If you want the authentication of deployment, you can visit https://astro-theme-pure-n7se13krf-cworld.vercel.app/ and send me the request (you know, Vercel is good for personal usage, but has some limitations for collaboration on free account)

Yep, I saw that it seems that the rendering order of nested components wouldn't compute the locale in the intended order, I pushed few more commits on that. 

Regarding the content not rendering content other than default, I assume you're talking about getting 404 error page when you enter pages other than blog and there's a decision to make that I mentioned before. I can think of three approaches: 

## Option A
Duplicate the same pages structure under [lang]. Right now it's set only for the blog page and the home. 
**Pro:** Internationalization would be ready to use and who doesn't want it, could just delete [lang] folder. 
**Con:** Basically duplication of the pages in the repo.

## Option B
Keep only pages under [lang].
**Pro:** Internationalization ready to use and no duplication.
**Con:** For those who don't want / need it and want to keep clean monolingual structure would have to create their own pages.

## Option C
We leave only the backbones of internationalization along with helper functions, ready to use components and no [lang] pages by default.
**Pro:** Better for those who won't include internationalization.
**Con:** For those who want internationalization would have to create their own pages under [lang]

**In summary:** It's deciding how much of the feature would be usable out of the box.

Until this point only archive, tags and sitemap is remaining to be ready for review, I'll keep working on those later.

Sorry for the long reply I think it's important to discuss it, I will try the vercel request, thanks!

agsoto commented:
Alright, there might be more options fixing the bug that was happening on `npm run build`, refactor on [6411929](https://github.com/cworld1/astro-theme-pure/pull/102/commits/6411929e332bf632721a6b88ddf1451843198ad3) provide routes for static output. Maybe there's chance for other options, but those would just automate the pages / folder structures duplication as variants of option A.

cworld1 commented:
Your code is starting to get bigger, which makes me confused about how you implement your features, that affects me to give you some indirect help that might be useful. It is better to read the official documentation carefully (I feel like I would get something new every time I read it), and the official example project [starlight](https://github.com/withastro/starlight)

agsoto committed:
refactor(i18n): use built in i18n implementations and migrate locale files to JSON format

- Replace TypeScript locale files with JSON format for better maintainability
- Remove locale redirect functionality and related components
- Simplify i18n implementation using Astro's built-in i18n utilities
- Update components to use getRelativeLocaleUrl instead of localizeLink
- Consolidate locale-related utility functions
- Add support for content-based i18n loading
- Improve language switch component with better accessibility

agsoto committed:
fix(i18n): Remove i18n missing folder warning

agsoto committed:
feat(i18n): add locale handling for tags, archives and search

- Replace 'language' field with 'lang' in blog posts for consistency
- Add getLanguageDisplayName utility for locale name formatting
- Update i18n JSON files with new structure and translations
- Implement locale-aware routing for tags and archives
- Refactor static path generation for localized pages
- Update UI components to use new i18n system

agsoto committed:
feat(i18n): add localized rss feed support with image handling

- Add new localized rss feed endpoint at /[lang]/rss.xml
- Refactor image path resolution to support both default and localized blog directories
- Improve hero image URL handling with external URL support
- Normalize blog post URLs for locale support

agsoto committed:
chore(i18n): add missing translations and link handling

- Add missing translations for blog and common strings in all supported languages
- Add left locale handling to use Astro.currentLocale consistently
- Refactor messy static path generation for localized routes
- Improve back button navigation with proper locale URLs

agsoto committed:
chore(rebase): resolve conflicts after rebasing to upstream v4.1.0

agsoto committed:
chore(i18n): add translations for theme change and link copied toasts

cworld1 commented:
plz consider rebase (if it is very hard for rebase, merge is also allowed) main branch with new coming commits. This will help us better accept your pr without a lot of conflicts. tk you again😉

cworld1 commented on tsconfig.json:
```diff
@@ -23,7 +23,16 @@
       "@/pages/*": ["src/pages/*"],
       "@/types": ["src/types/index.ts"],
-      "@/site-config": ["src/site.config.ts"]
+      "@/site-config": ["src/site.config.ts"],
```
you can consider https://astro-pure.js.org/docs/advanced/customize#package-mode turn to package mode development, this may help you better querying the needed package

cworld1 commented:
if you've done that work, you can tell me here. I will take the time to make better small changes

agsoto committed:
chore(rebase): resolve conflicts after rebasing to upstream v4.1.1

agsoto committed:
chore(rebase): resolve conflicts and issues after rebasing to upstream v4.1.2

agsoto commented:
@cworld1 I'm wondering if the changes are going to be merged next, as they supposed to be on v4.1.3? Otherwise I see no point on rebasing endlessly and we should rather drop and close the PR

cworld1 commented:
Sorry for that. I really like this feature in fact, but you know I haven't been able to find time recently, and some basic maintenance work still needs me to finish. In fact, v4.1.3 only makes basic maintenance and doesn't add any new substantial features.

cworld1 commented:
If you have patience, you can choose to close this PR first. And in the future, when I have time or interest, I will leave a message here inviting you to open a new collaboration. How about this?

agsoto commented:
Let's keep this open then. I hope when you can dedicate some effort to this you will tag me here again, expecting I'll also provide help if needed for the merge ✌️
